text/0112-server-engine.md

```diff
@@ -0,0 +1,558 @@
+
+Using the mocking approach, the existing `@lwc/engine` stays untouched and the entry point of the SSR version of LWC would look something like the following.
+
+**`server.ts`:**
+
+```ts
+// Evaluate the DOM polyfills before the DOM engine. Re-export all the exported properties from the
+// DOM engine.
+import './dom-environment';
+export * from '@lwc/engine';
+```
+
+**`dom-environment.ts`:**
+
+```ts
+class Node {
+    children: Node[] = [];
+
+    insertBefore(child: Node, anchor: Node): void {
+        const anchorIndex = this.children.indexOf(anchor);
+
+        if (anchorIndex === -1) {
+            this.children.push(child);
+        } else {
+            this.children.splice(anchorIndex, 0, child);
+        }
+    }
+}
+
+class Element extends Node {
+    tagName: string;
+
+    constructor(tagName: string) {
+        super();
+        this.tagName = tagName;
+    }
+}
+
+const document = {
+    createElement(tagName: string): Element {
+        return new Element(tagName);
+    },
+};
+
+const window = {
+    document,
+};
+
+// Assigning the synthetic DOM APIs to the current environment global object.
+Object.assign(globalThis, {
+    window,
+    document,
+    Node,
+    Element,
+});
+```
+
+#### Runtime injection of the rendering APIs
+
+The second approach is to **abstract the DOM APIs and the core UI framework**. This is what [React](https://github.com/facebook/react/tree/master/packages/react-dom), [Vue](https://github.com/vuejs/vue-next/tree/master/packages/runtime-dom), and [Angular](https://github.com/angular/angular/tree/master/packages/platform-browser/src) do. This involves an indirection between the DOM APIs and the core framework code. The indirection is injected at runtime depending on the environment. When loaded in the browser, the rendering APIs create DOM `Element` and set DOM `Attribute` on the elements. When loaded on the server, the rendering APIs manipulate strings to serialize components on the fly.
+
+When using a type-safe language like TypeScript, it is possible to ensure that all the rendering APIs are fulfilled in all environments at compile-time. In addition, injecting this indirection at runtime ensures that component code doesn't have access to actual DOM APIs when running on the server.
+
+From a drawback perspective, introducing this layer of indirection between the core logic and underlying APIs might introduce a performance overhead and increase the overall size of the engine.
+
+To inject the rendering APIs depending on the environment, we would need to create a different entry point per environment. Below you can find some pseudo-code for each entry point. The two entry points use the same `createComponent` abstraction to create an LWC component but pass different renderers specific to the target environment.
+
+**`entry-points/dom.js`:**
+
+```ts
+import { LightningElement, createComponent } from '@lwc/engine-core';
+
+const domRenderer = {
+    createElement(tagName: string): Element {
```

Comment:
super tiny: you have this as a file called `dom.js` but it is a typescript file :)